app/models/payment_intent.rb

```diff
@@ -0,0 +1,153 @@
+      update_stripe_status_and_charges(api_intent, action_source, source_datetime, &block)
+    else
+      raise "Trying to update status and charges for a PaymentIntent with unmatched payment_record_type of: #{payment_record_type}"
+    end
+  end
+
+  def update_stripe_status_and_charges(api_intent, action_source, source_datetime)
+    ActiveRecord::Base.transaction do
+      self.update!(error_details: api_intent.last_payment_error)
+      self.payment_record.assign_status(api_intent) # Will also update the wca_status
+
+      # Payment Intent lifecycle as per https://stripe.com/docs/payments/intents#intent-statuses
+      case api_intent.status
+      when 'succeeded'
+        # The payment didn't need any additional actions and is completed!
+
+        # Record the success timestamp if not already done
+        unless self.succeeded?
+          self.assign_attributes(
+            confirmed_at: source_datetime,
+            confirmation_source: action_source,
+            wca_status: payment_record.determine_wca_status,
+          )
+          save
+          payment_record.save
+        end
+
+        intent_charges = Stripe::Charge.list(
+          { payment_intent: self.stripe_id },
+          stripe_account: self.find_account_id,
+        )
+
+        intent_charges.data.each do |charge|
+          recorded_transaction = StripeRecord.find_by(stripe_id: charge.id)
+
+          if recorded_transaction.present?
+            recorded_transaction.update_status(charge)
+          else
+            fresh_transaction = StripeRecord.create_from_api(charge, {})
+            fresh_transaction.update!(parent_transaction: self.payment_record)
+
+            # Only trigger outer update blocks for charges that are actually successful. This is reasonable
+            # because we only ever trigger this block for PIs that are marked "successful" in the first place
+            charge_successful = fresh_transaction.stripe_status == "succeeded"
+
+            yield fresh_transaction if block_given? && charge_successful
+          end
+        end
+      when 'canceled'
+        # Canceled by Stripe
+        self.update!(
+          canceled_at: source_datetime,
+          cancellation_source: action_source,
+          wca_status: payment_record.determine_wca_status,
+        )
+        save
+        payment_record.save
+      when 'requires_payment_method'
+        # Reset by Stripe
+        self.update!(
+          confirmed_at: nil,
+          confirmation_source: nil,
+          canceled_at: nil,
+          cancellation_source: nil,
+          wca_status: payment_record.determine_wca_status,
+        )
+        save
+        payment_record.save
+      else
+        self.update!(wca_status: payment_record.determine_wca_status)
+        save
+        payment_record.save
+      end
+    end
+  end
+
+  # Sets the wca_status field based on the status field of the associated PaymentRecord
+  def assign_wca_status
```

Comment:
Where did the exclamation mark that I told you about in a previous review go?
`def assign_wca_status!`

Reply:
Removed because we are no longer persisting the updated state to the db, and thus are no longer "perform[ing] some permanent or potentially dangerous change" (looked it up on [SO](https://stackoverflow.com/questions/7179016/what-is-the-purpose-of-and-at-the-end-of-method-names) to figure out if I should keep the `!` or not

Reply:
Interesting tidbit from that thread btw: 
> One important note is that you should only have a bang method if you also have a corresponding non-bang method. The bang is used to distinguish the "more surprising" version of the method from the "less surprising" one. If you have only one method, then there's no need for a distinction, and you shouldn't name it with a bang. See Array#clear, for example. It clears the array. Clearing the array naturally mutates it. There's nothing surprising about that, the name already makes it clear, therefore: no bang. See [ruby-forum.com/topic/176830#773946](https://www.ruby-forum.com/topic/176830#773946)